### PR TITLE
[npr]: get stream url from json-ld

### PR DESCRIPTION
**Pull request by r5d**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [ ] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Extract JSON-LD from the NRP webpage if it's available. If the
"format" and "stream" fields in the multimedia dictionary are 
empty in the result returned by the NPR API, use the 
embedURL from the JSON-LD, if it exists, as a stream url.

Update `InfoExtractor._json_ld.traverse_json_ld` to look for a
VideoObject in "subjectOf" if the `item_type` is "NewsArticle".

Update `InfoExtractor._json_ld.extract_video_object` to extract
"embedUrl" from the VideoObject.

Addresses the issue downloading the [Bonobo Tiny Desk Concert][bonobo] (#1934)

[bonobo]: https://www.npr.org/2022/03/15/1084896560/bonobo-tiny-desk-home-concert

### CODE REVIEW

File: yt_dlp/extractor/common.py [hunk 1]
```diff
@@ -1503,6 +1503,7 @@ def extract_video_object(e):
             author = e.get('author')
             info.update({
                 'url': url_or_none(e.get('contentUrl')),
+                'embedUrl': url_or_none(e.get('embedUrl')),
```
Comment (pukkandan) [inline yt_dlp/extractor/common.py:1506]: info must obey extractor specifications. `embedURL` is not a defined field

File: yt_dlp/extractor/common.py [hunk 2]
```diff
@@ -1569,6 +1570,8 @@ def traverse_json_ld(json_ld, at_top_level=True):
                     })
                     if traverse_obj(e, ('video', 0, '@type')) == 'VideoObject':
                         extract_video_object(e['video'][0])
+                    if traverse_obj(e, ('subjectOf', 0, '@type')) == 'VideoObject':
```
Comment (pukkandan) [inline yt_dlp/extractor/common.py:1573]: elif?

File: yt_dlp/extractor/npr.py [hunk 3]
```diff
@@ -65,6 +65,14 @@ def _real_extract(self, url):
...
         playlist_title = story.get('title', {}).get('$text')
 
+        # Fetch the JSON-LD from the npr page.
+        json_ld = self._search_json_ld(
+            self._download_webpage(url, playlist_id),
+            playlist_id,
+            'NewsArticle',
+            fatal=False
+        )
+
```
Comment (pukkandan) [inline yt_dlp/extractor/npr.py:75]: read style guide

File: yt_dlp/extractor/npr.py [hunk 4]
```diff
@@ -110,6 +118,13 @@ def _real_extract(self, url):
                 formats.extend(self._extract_m3u8_formats(
                     stream_url, stream_id, 'mp4', 'm3u8_native',
                     m3u8_id='hls', fatal=False))
+
+            if len(formats) == 0 and json_ld.get('embedUrl', None) is not None:
+                # Add the embedUrl as a stream.
+                formats.extend(self._extract_m3u8_formats(
+                    json_ld.get('embedUrl'), 'hlsUrl', 'mp4', 'm3u8_native',
```
Comment (pukkandan) [inline yt_dlp/extractor/npr.py:125]:
```suggestion
                    json_ld.get('embedUrl'), media_id, 'mp4',
```

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/common.py (hunk 1 above)
Comment (r5d) [reply yt_dlp/extractor/common.py]: updated this. now injects `embedUrl` into the `formats` field instead.

File: yt_dlp/extractor/common.py (hunk 2 above)
Comment (r5d) [reply yt_dlp/extractor/common.py]: fixed it.

File: yt_dlp/extractor/npr.py (hunk 3 above)
Comment (r5d) [reply yt_dlp/extractor/npr.py]: fixed it.

File: yt_dlp/extractor/npr.py (hunk 4 above)
Comment (r5d) [reply yt_dlp/extractor/npr.py]: fixed it.

**Comment by pukkandan** [comment]
Add a test

File: yt_dlp/extractor/common.py [hunk 5]
```diff
@@ -1501,8 +1501,9 @@ def extract_chapter_information(e):
         def extract_video_object(e):
             assert e['@type'] == 'VideoObject'
             author = e.get('author')
+            embed_url = url_or_none(e.get('embedUrl'))
```
Comment (pukkandan) [inline yt_dlp/extractor/common.py:1504]:
```suggestion
```

**Comment by r5d** [comment]
@pukkandan added a test for the json_ld use case.

**Pull request merged**